pkg/rate/api_limiter.go

```diff
@@ -0,0 +1,790 @@
+	}
+
+	if l.params.SkipInitial > 0 && requestsProcessed < int64(l.params.SkipInitial) {
+		goto skipRateLimiter
+	}
+
+	if l.params.MinWaitDuration > 0 && limitWaitDuration < l.params.MinWaitDuration {
+		limitWaitDuration = l.params.MinWaitDuration
+	}
+
+	if l.params.MaxWaitDuration > 0 && limitWaitDuration > l.params.MaxWaitDuration {
+		if l.params.Log {
+			scopedLog.Warning("Not processing API request. Wait duration exceeds maximum")
+		}
+		return nil, fmt.Errorf("request would have to wait %v to be served (maximum wait duration: %v)",
+			limitWaitDuration, l.params.MaxWaitDuration)
+	}
+
+	if limitWaitDuration != 0 {
+		select {
+		case <-time.After(limitWaitDuration):
+		case <-ctx.Done():
+			if l.params.Log {
+				scopedLog.Warning("Not processing API request due to cancelled context while waiting")
+			}
+			return nil, fmt.Errorf("request cancelled while waiting for rate limiting slot: %w", ctx.Err())
+		}
+	}
+
+	if parallelRequests > 0 {
+		waitCtx := ctx
+		if l.params.MaxWaitDuration > 0 {
+			maxWaitDurationParallel := l.params.MaxWaitDuration - limitWaitDuration
+			scopedLog = scopedLog.WithField(logMaxWaitDurationParallel, maxWaitDurationParallel)
+			ctx2, cancel := context.WithTimeout(ctx, maxWaitDurationParallel)
+			defer cancel()
+			waitCtx = ctx2
+		}
+		waitSemaphoreWeight = int64(waitSemaphoreResolution / parallelRequests)
+		err := l.parallelWaitSemaphore.Acquire(waitCtx, waitSemaphoreWeight)
+		if err != nil {
+			if l.params.Log {
+				scopedLog.Warning("Not processing API request. Wait duration for maximum parallel requests exceeds maximum")
+			}
+			return nil, fmt.Errorf("timed out while waiting to be served with %d parallel requests: %w", parallelRequests, err)
+		}
+	}
+
+	waitDuration = time.Since(scheduledAt)
+
+skipRateLimiter:
+
+	l.mutex.Lock()
+	l.currentRequestsInFlight++
+	l.mutex.Unlock()
+
+	log.WithFields(logrus.Fields{
+		logAPICallName:       l.name,
+		logUUID:              uuid,
+		logWaitDurationTotal: waitDuration,
+	}).Debug("API call is ready to be served")
+
+	if l.params.Log {
+		scopedLog.Info("Processing API request with rate limiter")
+	} else {
+		scopedLog.Debug("Processing API request with rate limiter")
+	}
+
+	return &limitedRequest{
+		limiter:             l,
+		startTime:           time.Now(),
+		scheduleTime:        scheduledAt,
+		waitDuration:        waitDuration,
+		waitSemaphoreWeight: waitSemaphoreWeight,
+		uuid:                uuid,
+	}, nil
+}
+
+func parseRate(r string) (rate.Limit, error) {
+	tokens := strings.SplitN(r, "/", 2)
+	if len(tokens) != 2 {
+		return 0, fmt.Errorf("not in the form number/interval")
+	}
+
+	f, err := strconv.ParseFloat(tokens[0], 64)
+	if err != nil {
+		return 0, fmt.Errorf("unable to parse float '%s': %w", tokens[0], err)
+	}
+
+	// If duration is provided as "m" or "1", convert it into "1m" or "1s"
+	if _, err := strconv.ParseInt(string(tokens[1][0]), 10, 64); err != nil {
+		tokens[1] = "1" + tokens[1]
+	}
```

Comment:
I see the `m` / `s` -> `1m` / `1s` conversion here, but I don't see the interpretation of `1`. Seems like if we pass this down below then we'd interpret it as 1ns which I don't think makes any sense in context of the ratelimit package.

Reply:
The purpose of this is that you can write "5/m" instead of "5/1m". Maybe I don't understand your comment properly.

Reply:
My concern is that if the user specifies `5/1` then this will be something like "5 times per nanosecond".

Reply:
So you want to fail if `tokens[1]` can be parsed as an in if I understand you correctly?

Reply:
Changed